src/main/java/jenkins/plugins/git/GitSCMSource.java

```diff
@@ -102,81 +114,225 @@
+            traits.add(trait);
+        }
     }
 
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("3.4.0")
     public GitSCMSource(String id, String remote, String credentialsId, String includes, String excludes, boolean ignoreOnPushNotifications) {
         this(id, remote, credentialsId, null, null, includes, excludes, ignoreOnPushNotifications);
     }
 
-    public boolean isIgnoreOnPushNotifications() {
-      return ignoreOnPushNotifications;
+    private Object readResolve() throws ObjectStreamException {
+        if (traits == null) {
+            traits = new ArrayList<>();
+            if ((includes != null && !DEFAULT_INCLUDES.equals(includes))
+                    || (excludes != null && !DEFAULT_EXCLUDES.equals(excludes))) {
+                traits.add(new WildcardSCMHeadFilterTrait(includes, excludes));
+            }
+            if (extensions != null) {
+                EXTENSIONS:
+                for (GitSCMExtension extension : extensions) {
+                    for (SCMSourceTraitDescriptor d : SCMSourceTrait.all()) {
+                        if (d instanceof GitSCMExtensionTraitDescriptor) {
+                            GitSCMExtensionTraitDescriptor descriptor = (GitSCMExtensionTraitDescriptor) d;
+                            if (descriptor.getExtensionClass().isInstance(extension)) {
+                                try {
+                                    SCMSourceTrait trait = descriptor.convertToTrait(extension);
+                                    if (trait != null) {
+                                        traits.add(trait);
+                                        continue EXTENSIONS;
+                                    }
+                                } catch (UnsupportedOperationException e) {
+                                    LOGGER.log(Level.WARNING,
+                                            "Could not convert " + extension.getClass().getName() + " to a trait", e);
+                                }
+                            }
+                        }
+                        LOGGER.log(Level.FINE, "Could not convert {0} to a trait (likely because this option does not "
+                                + "make sense for a GitSCMSource)", getClass().getName());
+                    }
+                }
+            }
+            if (remoteName != null && !DEFAULT_REMOTE_NAME.equals(remoteName) && StringUtils.isNotBlank(remoteName)) {
+                traits.add(new RemoteNameSCMSourceTrait(remoteName));
+            }
+            if (StringUtils.isNotBlank(gitTool)) {
+                traits.add(new GitToolSCMSourceTrait(gitTool));
+            }
+            if (browser != null) {
+                traits.add(new GitBrowserSCMSourceTrait(browser));
+            }
+            RefSpecsSCMSourceTrait trait = asRefSpecsSCMSourceTrait(rawRefSpecs, remoteName);
+            if (trait != null) {
+                traits.add(trait);
+            }
```

Comment:
should probably set most fields to null after the conversion to potentially be able to collect some garbage.